### ISSUE TITLE
Postgres CTDL API search tuning

### ISSUE THREAD

**Issue by siuc-nate** [CredentialEngine/CredentialRegistry#471]
Yesterday, we tested with this query, which yielded 130 results:
```
{
	"@type": [
		"ceterms:ApprenticeshipCertificate",
		"ceterms:AssociateDegree",
		"ceterms:BachelorDegree",
		"ceterms:Badge",
		"ceterms:Certificate",
		"ceterms:CertificateOfCompletion",
		"ceterms:Certification",
		"ceterms:Degree",
		"ceterms:DigitalBadge",
		"ceterms:Diploma",
		"ceterms:DoctoralDegree",
		"ceterms:GeneralEducationDevelopment",
		"ceterms:JourneymanCertificate",
		"ceterms:License",
		"ceterms:MasterCertificate",
		"ceterms:MasterDegree",
		"ceterms:MicroCredential",
		"ceterms:OpenBadge",
		"ceterms:ProfessionalDoctorate",
		"ceterms:QualityAssuranceCredential",
		"ceterms:ResearchDoctorate",
		"ceterms:SecondarySchoolDiploma"
	],
	"ceterms:occupationType": {
		"ceterms:targetNodeDescription": "search:anyValue",
		"ceterms:codedNotation": {
			"search:value": "23-2011.00",
			"search:matchType": "search:exactMatch"
		},
		"search:operator": "search:orTerms"
	},
	"search:termGroup": {
		"search:operator": "search:orTerms",
		"ceterms:availableAt": {
			"ceterms:addressRegion": [
				{ "search:value": "indiana" },
				{ "search:value": "in", "search:matchType": "search:exactMatch" }
			]
		},
		"ceterms:offeredBy": {
			"ceterms:address": {
				"ceterms:addressRegion": [
					{ "search:value": "indiana" },
					{ "search:value": "in", "search:matchType": "search:exactMatch" }
				]
			}
		}
	}
}
```

However, the occupationType portion has some redundancies. When those are removed (so that we're just looking for the specific code), we get 3 results in the SPARQL API but 0 results in the Postgres API:
```
{
	"@type": [
		"ceterms:ApprenticeshipCertificate",
		"ceterms:AssociateDegree",
		"ceterms:BachelorDegree",
		"ceterms:Badge",
		"ceterms:Certificate",
		"ceterms:CertificateOfCompletion",
		"ceterms:Certification",
		"ceterms:Degree",
		"ceterms:DigitalBadge",
		"ceterms:Diploma",
		"ceterms:DoctoralDegree",
		"ceterms:GeneralEducationDevelopment",
		"ceterms:JourneymanCertificate",
		"ceterms:License",
		"ceterms:MasterCertificate",
		"ceterms:MasterDegree",
		"ceterms:MicroCredential",
		"ceterms:OpenBadge",
		"ceterms:ProfessionalDoctorate",
		"ceterms:QualityAssuranceCredential",
		"ceterms:ResearchDoctorate",
		"ceterms:SecondarySchoolDiploma"
	],
	"ceterms:occupationType": {
		"ceterms:codedNotation": {
			"search:value": "23-2011.00",
			"search:matchType": "search:exactMatch"
		}
	},
	"search:termGroup": {
		"search:operator": "search:orTerms",
		"ceterms:availableAt": {
			"ceterms:addressRegion": [
				{ "search:value": "indiana" },
				{ "search:value": "in", "search:matchType": "search:exactMatch" }
			]
		},
		"ceterms:offeredBy": {
			"ceterms:address": {
				"ceterms:addressRegion": [
					{ "search:value": "indiana" },
					{ "search:value": "in", "search:matchType": "search:exactMatch" }
				]
			}
		}
	}
}
```

Results like this one should be returned by the above query in both cases:
https://credentialengineregistry.org/resources/ce-414b8b21-80a6-4f92-918d-87b04f264556
(Note the occupationType that has a codedNotation value of exactly `23-2011.00`

Similarly, the same query but with a different code (`29-1141.00`) yields 17 results in SPARQL but 0 with the new search:
```
{
	"@type": [
		"ceterms:ApprenticeshipCertificate",
		"ceterms:AssociateDegree",
		"ceterms:BachelorDegree",
		"ceterms:Badge",
		"ceterms:Certificate",
		"ceterms:CertificateOfCompletion",
		"ceterms:Certification",
		"ceterms:Degree",
		"ceterms:DigitalBadge",
		"ceterms:Diploma",
		"ceterms:DoctoralDegree",
		"ceterms:GeneralEducationDevelopment",
		"ceterms:JourneymanCertificate",
		"ceterms:License",
		"ceterms:MasterCertificate",
		"ceterms:MasterDegree",
		"ceterms:MicroCredential",
		"ceterms:OpenBadge",
		"ceterms:ProfessionalDoctorate",
		"ceterms:QualityAssuranceCredential",
		"ceterms:ResearchDoctorate",
		"ceterms:SecondarySchoolDiploma"
	],
	"ceterms:occupationType": {
		"ceterms:codedNotation": {
			"search:value": "29-1141.00",
			"search:matchType": "search:exactMatch"
		}
	},
	"search:termGroup": {
		"search:operator": "search:orTerms",
		"ceterms:availableAt": {
			"ceterms:addressRegion": [
				{ "search:value": "indiana" },
				{ "search:value": "in", "search:matchType": "search:exactMatch" }
			]
		},
		"ceterms:offeredBy": {
			"ceterms:address": {
				"ceterms:addressRegion": [
					{ "search:value": "indiana" },
					{ "search:value": "in", "search:matchType": "search:exactMatch" }
				]
			}
		}
	}
}
```

Strangely, however, changing the code to just `23` and the matchType to `search:startsWith` yields what appears to be correct results in the new API but incorrect results (it includes non-matching false positives) in SPARQL (I think there is a bug in my translation code dealing with the matchType here)
```
{
	"@type": [
		"ceterms:ApprenticeshipCertificate",
		"ceterms:AssociateDegree",
		"ceterms:BachelorDegree",
		"ceterms:Badge",
		"ceterms:Certificate",
		"ceterms:CertificateOfCompletion",
		"ceterms:Certification",
		"ceterms:Degree",
		"ceterms:DigitalBadge",
		"ceterms:Diploma",
		"ceterms:DoctoralDegree",
		"ceterms:GeneralEducationDevelopment",
		"ceterms:JourneymanCertificate",
		"ceterms:License",
		"ceterms:MasterCertificate",
		"ceterms:MasterDegree",
		"ceterms:MicroCredential",
		"ceterms:OpenBadge",
		"ceterms:ProfessionalDoctorate",
		"ceterms:QualityAssuranceCredential",
		"ceterms:ResearchDoctorate",
		"ceterms:SecondarySchoolDiploma"
	],
	"ceterms:occupationType": {
		"ceterms:codedNotation": {
			"search:value": "23",
			"search:matchType": "search:startsWith"
		}
	},
	"search:termGroup": {
		"search:operator": "search:orTerms",
		"ceterms:availableAt": {
			"ceterms:addressRegion": [
				{ "search:value": "indiana" },
				{ "search:value": "in", "search:matchType": "search:exactMatch" }
			]
		},
		"ceterms:offeredBy": {
			"ceterms:address": {
				"ceterms:addressRegion": [
					{ "search:value": "indiana" },
					{ "search:value": "in", "search:matchType": "search:exactMatch" }
				]
			}
		}
	}
}
```

Is there perhaps a bug of some sort with the way codedNotation is being tokenized and/or something with the way exactMatch is handled?

**Comment by siuc-nate** [comment]
One other thing to note, that makes me think it might have to do with tokenization: I removed the "exact match" part of the query and tried just looking for a codedNotation of "23-2" (as well as "23-2011", "23 2011", "23 2", etc). This gives 0 results. The only way to get results with the above query is to use just "23" as the code.

**Comment by siuc-nate** [comment]
In fact, cutting the query down to just the code yields 0 results:
![image](https://user-images.githubusercontent.com/21346519/127685569-10cd185c-face-4b99-9c08-fa2555710ee9.png)

**Comment by excelsior** [comment]
Thank for the feedback, Nate!

You're right, search term sanitization wasn't done right: the column was sanitized using the built-in PG function, but the term itself was sanitized in the Ruby code, so the two things were not on the same page. I fixed that problem and a coulpe of smaller bugs I found in the process.

A few words about the match type thing. A value like `23-2011.00` can't be stemmed, meaning that it always gets matched as is (using the `simple` dictionary), the `search:matchType` operator gets ignored in this case.

The `search:startsWith` match type isn't supported yet. I don't think it's possible to implement it by the means of PG's FTS, we'll have to use `LIKE` for such conditions (it'd require a different index).

I'm studying the C# code you sent me in order to find out what else is missing in my implementation.

**Comment by siuc-nate** [comment]
I think I may have come across another issue - reverse connections.  The current search API defaults to looking in both directions for every property which references an object (unless the user explicitly searches in just one direction).  This allows fairly important use cases like "show me all of the competency frameworks with a competency that says 'skills'".  

For instance:
```
{
	"ceasn:isPartOf": [
		{
			"ceasn:competencyText": "skills"
		}
	]
}
```
generates this SPARQL:
```
PREFIX credreg: <https://credreg.net/> PREFIX ceasn: <https://purl.org/ctdlasn/terms/> PREFIX ceterms: <https://purl.org/ctdl/terms/> SELECT ?totalResults ?id ?searchResultPayload  WITH { SELECT DISTINCT ?id ?searchResultPayload (SUM(COALESCE(?relevance_1, 0)) AS ?sortOrder_0)  WHERE { <http://aws.amazon.com/neptune/vocab/v01/QueryHints#Query> <http://aws.amazon.com/neptune/vocab/v01/QueryHints#joinOrder> 'Ordered' . FILTER EXISTS { ?id ceterms:ctid ?anyValue . } { ?id ( (ceasn:isPartOf)|(^ceasn:isPartOf) ) ?layer_1 { ?layer_1 ( ceasn:competencyText__tokenData ) ?layer_2 { { ?layer_2 ( credreg:__tokenFullNormalized ) ?token_1 . BIND(IF(REGEX(?token_1, 'skills'), 1500, IF(REGEX(?token_1, 'skill'), 150, IF(REGEX(?token_1, 'kill'), 4, 0))) AS ?relevance_1) FILTER(?relevance_1 > 0) . } } } } ?id credreg:__payload ?searchResultPayload .   } GROUP BY ?id ?searchResultPayload  } AS %mainQuery WHERE { { SELECT (COUNT(DISTINCT(?id)) AS ?totalResults) WHERE { INCLUDE %mainQuery } } UNION { SELECT ?id ?searchResultPayload ?sortOrder_0  WHERE { INCLUDE %mainQuery } ORDER BY DESC(?sortOrder_0) OFFSET 0 LIMIT 5 } } ORDER BY DESC(?sortOrder_0)
```
The key bit being that `id ( (ceasn:isPartOf)|(^ceasn:isPartOf) ) ?layer_1` part.

Is there a way to support reverse connection searching in the Postgres API? Preferably by default, since that's how the current code operates, but if needed we could require some extra flag for it (the equivalent of SPARQL's `^`).  Note that this reverse search only happens for properties that reference objects, not properties that end in text/numbers/booleans/etc.

**Comment by siuc-nate** [comment]
Per our 8/12/2021 meeting:

- The search will continue as-is now by default (forward-looking references)
- We will only do a reverse lookup when the user includes a `^` as the first character of a property

The above query would therefore look like this:
```
{
	"@type": "ceasn:CompetencyFramework",
	"^ceasn:isPartOf": [
		{
			"ceasn:competencyText": "some skill name"
		}
	]
}
```

**Comment by siuc-nate** [comment]
I was working on updating the finder's competency framework search to use the new API and I stumbled across another issue:
![image](https://user-images.githubusercontent.com/21346519/130509914-a49e618b-fc89-4e9a-91ab-f6595b7ea311.png)

`ceasn:comment` is a valid property:
https://credreg.net/ctdlasn/terms/comment

And it appears in the [context file](https://credreg.net/ctdl/schema/context/json):
![image](https://user-images.githubusercontent.com/21346519/130510155-9a6e7f51-bcb9-4254-97cc-8491a7e0656b.png)

So I'm not sure what the problem is. Any ideas?

**Comment by siuc-nate** [comment]
Ditto with this one.  Is sandbox not updated by chance?
![image](https://user-images.githubusercontent.com/21346519/130513214-bacf1ddf-00cb-4540-a833-f0a2b609ff63.png)

**Comment by excelsior** [comment]
@siuc-nate There are two separate issues here:

1. The table containing 'flattened' resources with indices and stuff gets built dynamically from the JSON-LD data, meaning that if a property doesn't exists, there will be no column for it. In other words, the context isn't the ultimate source of truth, the data itself is. In this particular case, there's no single resource with a `ceasn:comment` property on sandbox, that's why it's not possible to use it in a query:

```sql
SELECT COUNT(*) FROM "envelope_resources" WHERE (processed_resource ? 'ceasn:comment') -- returns nothing
```

I guess, you'd rather have it behave the way SPARQL does, i.e., returning nothing when a property doesn't exists instead of throwing an error?

2. We cache the context files for better performance, but this job hasn't been set up on sandbox, so the cache was out of date. I configured it be sync daily and am rebuilding the sandbox index, will ping you once it's ready.

**Comment by excelsior** [comment]
UPD: The index has been rebuilt, the latter query and the likes should work now.

**Comment by siuc-nate** [comment]
Thanks.  The search should minimally allow for any property that's in the schema (+ the special "search" properties), but the ideal behavior would be for it to either ignore things it doesn't understand or to, as you suggested, return nothing if there is no match. In this case, the `ceasn:comment` was part of an `OR`, so it should still be possible to return results based on the other properties even with that "invalid" property in the query.

In other words, treat invalid properties as simply not matching anything, so that the rest of the query's logic can still be applied as the user would likely expect.  Hopefully I'm making sense.

**Comment by excelsior** [comment]
@siuc-nate Oops, the property existency check was treating references the wrong way. Fixed that!

**Comment by siuc-nate** [comment]
Thanks, it gets results now.

However, I discovered another wrinkle with the `"search:matchType": "search:exactMatch"` feature.  This query:
```
{
	"^ceasn:isPartOf": {
		"ceasn:competencyText": {
			"search:value": "programming",
			"search:matchType": "search:exactMatch"
		}
	}
}
```
Some of the results it finds do have competencies with "programming" in their competencyText field, but some only have "program", leading to some false positives.

**Comment by excelsior** [comment]
My bad, the `search:matchType` option wasn't supported for language maps, only for plain strings. It should work now.

**Comment by siuc-nate** [comment]
Thanks, that one works now.

I tried testing the matchType with URL strings and am getting an error though:
```
{
	"ceasn:derivedFrom": {
		"search:value": "https://www.isc2.org/certifications/csslp",
		"search:matchType": "search:exactMatch"
	}
}
```
Should find https://credentialengineregistry.org/resources/ce-20a93efb-a485-4e0b-8747-529ac4c5e0e6

![image](https://user-images.githubusercontent.com/21346519/131520812-24b97144-932a-48b5-9516-14dd10de4f1c.png)

**Comment by excelsior** [comment]
Unfortunately, match types are not supported in subresource conditions like this one. I'll fix that, but for now you can query it in following way:

```json
{
  "ceasn:derivedFrom": "https://www.isc2.org/certifications/csslp"
}
```

**Comment by siuc-nate** [comment]
It looks like there's a bug with `search:anyValue`. The following query:
```
{
	"ceasn:source": "search:anyValue"
}
```
returns over 100 results with SPARQL and 0 with Postgres.

**Comment by excelsior** [comment]
The problem is, `ceasn:source` isn't truly a reference. Its typical values looks like this: http://nocti.org/PDFs/blueprint/1387.pdf — it's just a URL, it doesn't define a resource in the registry. The current implementation assumes there is a resource, even in the case of a `search:anyValue` condition. I guess I need to make some exceptions for cases like this one. I'll keep you posted.

**Comment by excelsior** [comment]
@siuc-nate `search:anyValue` is now works as supposed for `ceasn:source` and similar properties.

**Comment by siuc-nate** [comment]
Thanks.

How is sort order controlled with the new API? Currently with the SPARQL API the user can pass a "Sort" parameter that takes a few possible values (it should be documented in the SPARQL services file I sent a little while back).  I have also explored enabling more complex sort orders within the query itself (to allow for things like "order by x, then by y, then by z" and such) but I'm not sure how much of that made it into the new API.

**Comment by excelsior** [comment]
This option isn't available in the new API. I'll implement it ASAP.

**Comment by siuc-nate** [comment]
Thanks.  For the moment we just need to support a single-value sort; we may not need the multi-step sorting at all.

**Comment by siuc-nate** [comment]
One more thing I forgot about - being able to reverse the sort order.

Within the SPARQL implementation, I borrowed the idea of using `^` to indicate a reversal, so:
`Sort: "search:recordUpdated"` would return newest first and `Sort: "^search:recordUpdated"` would return oldest first.

Ditto for `search:recordCreated` and `search:relevance`, and for custom properties as well.  

For example:
`Sort: "ceterms:name"` would sort A-Z and `Sort: "^ceterms:name"` would sort Z-A (though I think I have those backwards in the current SPARQL implementation)

Can that be added?  Sorry for missing this one until just now.

**Comment by excelsior** [comment]
@siuc-nate Sort order is now can be controlled with the `order_by` query parameter. It accepts either a payload's top-level property or one of the special values: `search:recordCreated`, `search:recordUpdated`, and `search:relevance`.

Examples:

```
POST /ctdl?order_by=@type — A-Z
POST /ctdl?order_by=^ceterms:name — Z-A
POST /ctdl?order_by=search:recordUpdated — least recently updated
POST /ctdl?order_by=^search:recordCreated — most recently created
```

A value provided can be prefixed with `^`, which will reverse the order (it's ascending by default).

When no value is provided, the results are sorted by `^search:relevance` if there is an FTS condition or by `search:recordCreated` otherwise. Providing a non-supported value is equivalent to providing no value at all (no errors are raised in this case).

Specifying `search:relevance` for a query with no FTS conditions will fall back to `search: recordCreated`.

**Comment by siuc-nate** [comment]
Similarly, all four of these yield the same result:
```
https://credentialengineregistry.org/ctdl?skip=0&take=5&order_by=search:relevance
https://credentialengineregistry.org/ctdl?skip=0&take=5&order_by=^search:relevance
https://credentialengineregistry.org/ctdl?skip=0&take=5&order_by=%5Esearch%3Arelevance
https://credentialengineregistry.org/ctdl?skip=0&take=5&order_by=search%3Arelevance
```
with:
```
{
	 "@type": "ceasn:CompetencyFramework",
	 "ceasn:name": "nursing"
}
```
![image](https://user-images.githubusercontent.com/21346519/136467845-33abaf33-8f2c-4785-b067-b720d0f52a2a.png)

**Comment by siuc-nate** [comment]
The `search:recordCreated` and `search:recordUpdated` properties (when used as a filter within the query rather than as a sort order) don't seem to be working. Example:

All Certificate records created on/after 2021-01-01
```
{
	"@type": "ceterms:Certificate",
	"search:recordCreated": [ "2021-01-01", "" ]
}
```

All Certificate records created on/before 2021-01-01
```
{
	"@type": "ceterms:Certificate",
	"search:recordCreated": [ "", "2021-01-01" ]
}
```

All certificate records created between 2020-01-01 and 2021-01-01 (inclusive)
```
{
	"@type": "ceterms:Certificate",
	"search:recordCreated": [ "2020-01-01", "2021-01-01" ]
}
```

These all work in the SPARQL search but not the Postgres search.

**Comment by excelsior** [comment]
@siuc-nate The `search:recordCreated` and `search:recordUpdated` properties are now supported in the search API. Should the owner and publisher properties be also supported? If yes, how would you call them?

**Comment by siuc-nate** [comment]
If it's possible, I would say yes. Is the data stored as a URI or just a CTID (or some other way)? I could see something like this working (find all competency frameworks published by one of the organizations in the list):
```
{
  "@type": "ceterms:CompetencyFramework",
  "search:recordPublishedBy": [
    "ce-abcdef...",
    "ce-ghijk..."
    "ce-etcetc..."
  ]
}
```

Ideally it would work like any other property in CTDL.

**Comment by siuc-nate** [comment]
How do we indicate which registry community to search in? We have some special cases where we need to search (and get description set data, and so on) within a specific community.  Would something like one of these work?

```
https://credentialengineregistry.org/ctdl?skip=0&take=5&community=snhupr
https://credentialengineregistry.org/snhupr/ctdl?skip=0&take=5
```
Or does this already exist some other way?

Also, how does the description set system work within a private community? How would you access the description set data for something that exists in a private community? I assume you'd just call it the usual way with an appropriate key in the headers - but would you use the same `/description_sets` endpoint regardless of which community the data was in? Does the indexing process that builds the description set work across communities? I'm not entirely sure whether it should or not, given the security constraints on some communities. Maybe the description set for a resource in a private community can include data from public communities, but a description set for something in a public community can't include anything from a private community? @mparsons-ce what are your thoughts on this?

**Comment by excelsior** [comment]
@siuc-nate Nope, scoping by a community isn't supported in either API. I'm working on fixing that.

Separating private data from public data is a different task, let's open a new issue for that.

**Comment by edgarf** [comment]
(We should close this task and create separate ones for remaining issues @excelsior  @siuc-nate )

**Comment by siuc-nate** [comment]
Thanks @excelsior 

Closing this thread, as the issues have been resolved. If more come up, I'll give them their own threads.
